src/main/java/com/amazon/dlic/auth/http/jwt/keybyoidc/HTTPOpenIdAuthenticator.java

```diff
@@ -0,0 +1,289 @@
+
+    public HTTPJwtKeyByOpenIdConnectAuthenticator getOpenIdJwtAuthenticator() {
+        return openIdJwtAuthenticator;
+    }
+
+    public String getType() {
+        return "oidc";
+    }
+
+    private static SettingsBasedSSLConfigurator.SSLConfig getSSLConfig(Settings settings, Path configPath) throws Exception {
+        return new SettingsBasedSSLConfigurator(settings, configPath, "openid_connect_idp").buildSSLConfig();
+    }
+
+    @Override
+    public AuthCredentials extractCredentials(SecurityRequest request, ThreadContext context) throws OpenSearchSecurityException {
+        if (this.userInfoEndpoint != null && !this.userInfoEndpoint.isBlank()) {
+            return extractCredentials0(request, context);
+        }
+        return (this.openIdJwtAuthenticator.extractCredentials(request, context));
+    }
+
+    @Override
+    public Optional<SecurityResponse> reRequestAuthentication(SecurityRequest request, AuthCredentials credentials) {
+        return Optional.empty();
+    }
+
+    /**
+     * This method performs the logic required for making use of the userinfo_endpoint OIDC feature.
+     * Per the spec: https://openid.net/specs/openid-connect-core-1_0.html#UserInfo there are 10 verification steps we must perform
+     * 1. Validate the OP is correct via TLS certificate check
+     * 2. Validate response is OK
+     * 3. Validate application type is either JSON or JWT
+     * 4. Validate response is one of: just signed, just encrypted, or signed then encrypted (encryption MUST NOT occur before signing)
+     * 5. If response is signed then validate the signature via JWS and confirm the response has "iss" and "aud" claims
+     * 6. Validate "iss" claim is equal to the issuer ID url
+     * 7. Validate the "aud" claim is equal to the client ID
+     * 8. If the client provides a userinfo_encrypted_response_alg value decrypt the response using the keys from registration
+     * 9. Validate "sub" claim is always present
+     * 10. Validate "sub" claim matches the ID token
+     * @param request The SecurityRequest to perform auth on
+     * @param context The active thread context
+     * @return AuthCredentials formed through querying the userinfo_endpoint
+     * @throws OpenSearchSecurityException On failure to extract credentials from the request
+     */
+    public AuthCredentials extractCredentials0(SecurityRequest request, ThreadContext context) throws OpenSearchSecurityException {
+
+        try {
+
+            URI userInfoEndpointURI = new URI(this.userInfoEndpoint);
+
+            String bearerHeader = request.getHeaders().get(AUTHORIZATION).getFirst();
+            if (!StringUtils.isBlank(bearerHeader)) {
+                if (bearerHeader.contains("Bearer ")) {
+                    bearerHeader = bearerHeader.substring(7);
+                }
+            }
+
+            String finalBearerHeader = bearerHeader;
+
+            AccessToken accessToken = new AccessToken(AccessTokenType.BEARER, finalBearerHeader) {
```

Comment:
FYI Nimbus has a class called [BearerAccessToken](https://www.javadoc.io/doc/com.nimbusds/oauth2-oidc-sdk/5.9/com/nimbusds/oauth2/sdk/token/BearerAccessToken.html), would that be usable here?

Reply:
Looks like this is the source code: https://bitbucket.org/connect2id/oauth-2.0-sdk-with-openid-connect-extensions/src/c7736eb688bf0a81714374822d4e7b5232222eea/src/main/java/com/nimbusds/oauth2/sdk/token/BearerAccessToken.java#lines-208:211

Reply:
Oh I didn't realize that.